File: lib/sidereon/bodies.ex
defmodule Sidereon.Bodies do
  @moduledoc """
  Ground-observer Sun and Moon geometry.

  From a ground station and a UTC instant, answer the "observe the sky from a
  site" questions: the topocentric azimuth/elevation/range of the Sun or Moon,
  the Moon's illuminated fraction, and the Moon's rise/set and meridian-transit
  events over a window. The azimuth/elevation convention matches the satellite
  look-angle path; the Moon vector is geocentric, so the station-to-target
  reduction applies the diurnal parallax that matters for the nearby Moon.

  Precision follows the underlying analytic Sun/Moon series (sub-degree
  positions); this is a planning/visualization lens, not an almanac-grade
  reduction. The ephemeris, topocentric reduction, phase-angle geometry, and
  event refinement all live in the `sidereon-core` Rust core.

  A `station` is `{latitude_deg, longitude_deg, altitude_km}` or a map with
  `:latitude_deg`, `:longitude_deg`, `:altitude_km`. A time is a `DateTime`, a
  `NaiveDateTime` (interpreted as UTC), or a `{{y, m, d}, {h, min, s, us}}`
  tuple.
  """

  alias Sidereon.NIF

  @type station ::
          {number(), number(), number()}
          | %{latitude_deg: number(), longitude_deg: number(), altitude_km: number()}

  @type time :: DateTime.t() | NaiveDateTime.t() | {tuple(), tuple()}

  @typedoc "Topocentric look angle of a body from a site."
  @type az_el :: %{azimuth_deg: float(), elevation_deg: float(), range_km: float()}

  @typedoc "A Moon rise/set or meridian-transit event."
  @type event :: %{time: DateTime.t(), kind: atom(), elevation_deg: float()}

  @doc """
  Topocentric azimuth/elevation/range of the Sun from a site.
  """
  @spec sun_az_el(station(), time()) :: {:ok, az_el()} | {:error, atom()}
  def sun_az_el(station, time) do
    {lat, lon, alt} = station_parts(station)
    NIF.bodies_sun_az_el(lat, lon, alt, to_nif_datetime(time)) |> decode_az_el()
  end

  @doc """
  Topocentric azimuth/elevation/range of the Moon from a site.
  """
  @spec moon_az_el(station(), time()) :: {:ok, az_el()} | {:error, atom()}
  def moon_az_el(station, time) do
    {lat, lon, alt} = station_parts(station)
    NIF.bodies_moon_az_el(lat, lon, alt, to_nif_datetime(time)) |> decode_az_el()
  end

  @doc """
  Illuminated fraction of the Moon as seen from a site.

  Returns `{:ok, %{illuminated_fraction: k, phase_angle_deg: a}}` with `k` in
  `[0, 1]` (0 = new, 1 = full) and the Sun-Moon-observer phase angle in degrees.
  """
  @spec moon_illumination(station(), time()) ::
          {:ok, %{illuminated_fraction: float(), phase_angle_deg: float()}} | {:error, atom()}
  def moon_illumination(station, time) do
    {lat, lon, alt} = station_parts(station)

    case NIF.bodies_moon_illumination(lat, lon, alt, to_nif_datetime(time)) do
      {:ok, {fraction, phase_angle_deg}} ->
        {:ok, %{illuminated_fraction: fraction, phase_angle_deg: phase_angle_deg}}

      {:error, _reason} = err ->
        err
    end
  end

  @doc """
  Topocentric geometric Moon (disk-center) elevation at a site, degrees.
  """
  @spec moon_elevation_deg(station(), time()) :: {:ok, float()} | {:error, atom()}
  def moon_elevation_deg(station, time) do
    {lat, lon, alt} = station_parts(station)
    NIF.bodies_moon_elevation_deg(lat, lon, alt, to_nif_datetime(time))
  end

  @doc """
  Moon elevation-threshold crossings (moonrise / moonset) over a UTC window.

  ## Options

    * `:elevation_threshold_deg` - crossing threshold (default `-0.833`, the
      standard upper-limb-on-the-horizon convention).
    * `:step_seconds` - event-finder scan step (default `600.0`).
    * `:time_tolerance_seconds` - crossing-time refinement tolerance (default
      `1.0`).

  Returns `{:ok, [%{time: DateTime, kind: :rising | :setting, elevation_deg:}]}`.
  """
  @spec find_moon_elevation_crossings(station(), time(), time(), keyword()) ::
          {:ok, [event()]} | {:error, atom()}
  def find_moon_elevation_crossings(station, start_time, end_time, opts \\ []) do
    {lat, lon, alt} = station_parts(station)

    NIF.bodies_find_moon_elevation_crossings(
      lat,
      lon,
      alt,
      to_nif_datetime(start_time),
      to_nif_datetime(end_time),
      Keyword.get(opts, :elevation_threshold_deg, -0.833) / 1.0,
      Keyword.get(opts, :step_seconds, 600.0) / 1.0,
      Keyword.get(opts, :time_tolerance_seconds, 1.0) / 1.0
    )
    |> decode_events()
  end

  @doc """
  Moon meridian transits (upper and lower culminations) over a UTC window.

  ## Options

    * `:step_seconds` - event-finder scan step (default `600.0`).
    * `:time_tolerance_seconds` - crossing-time refinement tolerance (default
      `1.0`).

  Returns `{:ok, [%{time: DateTime, kind: :upper | :lower, elevation_deg:}]}`.
  """
  @spec find_moon_transits(station(), time(), time(), keyword()) ::
          {:ok, [event()]} | {:error, atom()}
  def find_moon_transits(station, start_time, end_time, opts \\ []) do
    {lat, lon, alt} = station_parts(station)

    NIF.bodies_find_moon_transits(
      lat,
      lon,
      alt,
      to_nif_datetime(start_time),
      to_nif_datetime(end_time),
      Keyword.get(opts, :step_seconds, 600.0) / 1.0,
      Keyword.get(opts, :time_tolerance_seconds, 1.0) / 1.0
    )
    |> decode_events()
  end

  # --- marshalling ----------------------------------------------------------

  defp station_parts({lat, lon, alt}) when is_number(lat) and is_number(lon) and is_number(alt),
    do: {lat / 1.0, lon / 1.0, alt / 1.0}

  defp station_parts(%{latitude_deg: lat, longitude_deg: lon, altitude_km: alt}), do: {lat / 1.0, lon / 1.0, alt / 1.0}

  defp to_nif_datetime(%DateTime{} = dt),
    do: {{dt.year, dt.month, dt.day}, {dt.hour, dt.minute, dt.second, elem(dt.microsecond, 0)}}

  defp to_nif_datetime(%NaiveDateTime{} = dt),
    do: {{dt.year, dt.month, dt.day}, {dt.hour, dt.minute, dt.second, elem(dt.microsecond, 0)}}

  defp to_nif_datetime({{_y, _m, _d}, {_h, _min, _s, _us}} = tuple), do: tuple

  defp decode_az_el({:ok, {azimuth_deg, elevation_deg, range_km}}),
    do: {:ok, %{azimuth_deg: azimuth_deg, elevation_deg: elevation_deg, range_km: range_km}}

  defp decode_az_el({:error, _reason} = err), do: err

  defp decode_events({:ok, rows}) do
    {:ok,
     Enum.map(rows, fn {unix_us, kind, elevation_deg} ->
       %{time: DateTime.from_unix!(unix_us, :microsecond), kind: kind, elevation_deg: elevation_deg}
     end)}
  end

  defp decode_events({:error, _reason} = err), do: err
end

File: native/sidereon_nif/src/bodies.rs
//! Rustler boundary for ground-observer Sun/Moon geometry.
//!
//! Pure glue over `sidereon_core::astro::bodies`: it decodes a ground station
//! (geodetic degrees / km altitude) and UTC datetime tuples, calls the crate's
//! observe/rise-set helpers, and encodes the look-angle, illumination, and
//! event results. The ephemeris, topocentric reduction, phase-angle geometry,
//! and event refinement all live in the core; no astronomy lives here.

use rustler::{Encoder, Env, Term};
use sidereon_core::astro::bodies::observe::{moon_az_el, moon_illumination, sun_az_el, BodyAzEl};
use sidereon_core::astro::bodies::rise_set::{
    find_moon_elevation_crossings, find_moon_transits, MoonElevationCrossing,
    MoonElevationCrossingKind, MoonElevationOptions, MoonTransit, MoonTransitKind,
};
use sidereon_core::astro::frames::transforms::GeodeticStationKm;
use sidereon_core::astro::passes::UtcInstant;

type DateTuple = (i32, i32, i32);
type TimeTuple = (i32, i32, i32, i32);
type AzElTerm = (f64, f64, f64);
// (unix_microseconds, kind_atom, elevation_deg)
type EventTerm<'a> = (i64, Term<'a>, f64);

mod atoms {
    rustler::atoms! {
        ok,
        error,
        invalid_input,
        rising,
        setting,
        upper,
        lower
    }
}

fn station(lat_deg: f64, lon_deg: f64, alt_km: f64) -> GeodeticStationKm {
    GeodeticStationKm {
        latitude_deg: lat_deg,
        longitude_deg: lon_deg,
        altitude_km: alt_km,
    }
}

/// Decode an Elixir `{{y,m,d},{h,min,s,us}}` UTC datetime into a [`UtcInstant`].
fn instant(datetime: Term) -> Option<UtcInstant> {
    let ((year, month, day), (hour, minute, second, microsecond)): (DateTuple, TimeTuple) =
        datetime.decode().ok()?;
    UtcInstant::from_utc(year, month, day, hour, minute, second, microsecond)
}

fn az_el_term(body: BodyAzEl) -> AzElTerm {
    (body.azimuth_deg, body.elevation_deg, body.range_km)
}

#[rustler::nif]
fn bodies_sun_az_el<'a>(
    env: Env<'a>,
    lat_deg: f64,
    lon_deg: f64,
    alt_km: f64,
    datetime: Term<'a>,
) -> Term<'a> {
    let Some(time) = instant(datetime) else {
        return (atoms::error(), atoms::invalid_input()).encode(env);
    };
    match sun_az_el(&station(lat_deg, lon_deg, alt_km), time) {
        Ok(body) => (atoms::ok(), az_el_term(body)).encode(env),
        Err(_) => (atoms::error(), atoms::invalid_input()).encode(env),
    }
}

#[rustler::nif]
fn bodies_moon_az_el<'a>(
    env: Env<'a>,
    lat_deg: f64,
    lon_deg: f64,
    alt_km: f64,
    datetime: Term<'a>,
) -> Term<'a> {
    let Some(time) = instant(datetime) else {
        return (atoms::error(), atoms::invalid_input()).encode(env);
    };
    match moon_az_el(&station(lat_deg, lon_deg, alt_km), time) {
        Ok(body) => (atoms::ok(), az_el_term(body)).encode(env),
        Err(_) => (atoms::error(), atoms::invalid_input()).encode(env),
    }
}

#[rustler::nif]
fn bodies_moon_illumination<'a>(
    env: Env<'a>,
    lat_deg: f64,
    lon_deg: f64,
    alt_km: f64,
    datetime: Term<'a>,
) -> Term<'a> {
    let Some(time) = instant(datetime) else {
        return (atoms::error(), atoms::invalid_input()).encode(env);
    };
    match moon_illumination(&station(lat_deg, lon_deg, alt_km), time) {
        Ok(illum) => (
            atoms::ok(),
            (illum.illuminated_fraction, illum.phase_angle_deg),
        )
            .encode(env),
        Err(_) => (atoms::error(), atoms::invalid_input()).encode(env),
    }
}

#[rustler::nif]
fn bodies_moon_elevation_deg<'a>(
    env: Env<'a>,
    lat_deg: f64,
    lon_deg: f64,
    alt_km: f64,
    datetime: Term<'a>,
) -> Term<'a> {
    let Some(time) = instant(datetime) else {
        return (atoms::error(), atoms::invalid_input()).encode(env);
    };
    // Route through `moon_az_el` (not the core `moon_elevation_deg`, which
    // `expect`s and would panic across the NIF on an invalid-but-finite station)
    // so a bad station surfaces as a typed `{:error, :invalid_input}`.
    match moon_az_el(&station(lat_deg, lon_deg, alt_km), time) {
        Ok(body) => (atoms::ok(), body.elevation_deg).encode(env),
        Err(_) => (atoms::error(), atoms::invalid_input()).encode(env),
    }
}

fn crossing_term<'a>(env: Env<'a>, crossing: MoonElevationCrossing) -> EventTerm<'a> {
    let kind = match crossing.kind {
        MoonElevationCrossingKind::Rising => atoms::rising().encode(env),
        MoonElevationCrossingKind::Setting => atoms::setting().encode(env),
    };
    (
        crossing.time.unix_microseconds(),
        kind,
        crossing.elevation_deg,
    )
}

fn transit_term<'a>(env: Env<'a>, transit: MoonTransit) -> EventTerm<'a> {
    let kind = match transit.kind {
        MoonTransitKind::Upper => atoms::upper().encode(env),
        MoonTransitKind::Lower => atoms::lower().encode(env),
    };
    (transit.time.unix_microseconds(), kind, transit.elevation_deg)
}

#[rustler::nif(schedule = "DirtyCpu")]
#[allow(clippy::too_many_arguments)]
fn bodies_find_moon_elevation_crossings<'a>(
    env: Env<'a>,
    lat_deg: f64,
    lon_deg: f64,
    alt_km: f64,
    start_datetime: Term<'a>,
    end_datetime: Term<'a>,
    elevation_threshold_deg: f64,
    step_seconds: f64,
    time_tolerance_seconds: f64,
) -> Term<'a> {
    let (Some(start), Some(end)) = (instant(start_datetime), instant(end_datetime)) else {
        return (atoms::error(), atoms::invalid_input()).encode(env);
    };
    let options = MoonElevationOptions {
        elevation_threshold_deg,
        step_seconds,
        time_tolerance_seconds,
    };
    match find_moon_elevation_crossings(&station(lat_deg, lon_deg, alt_km), start, end, options) {
        Ok(crossings) => {
            let rows: Vec<EventTerm> = crossings
                .into_iter()
                .map(|crossing| crossing_term(env, crossing))
                .collect();
            (atoms::ok(), rows).encode(env)
        }
        Err(_) => (atoms::error(), atoms::invalid_input()).encode(env),
    }
}

#[rustler::nif(schedule = "DirtyCpu")]
#[allow(clippy::too_many_arguments)]
fn bodies_find_moon_transits<'a>(
    env: Env<'a>,
    lat_deg: f64,
    lon_deg: f64,
    alt_km: f64,
    start_datetime: Term<'a>,
    end_datetime: Term<'a>,
    step_seconds: f64,
    time_tolerance_seconds: f64,
) -> Term<'a> {
    let (Some(start), Some(end)) = (instant(start_datetime), instant(end_datetime)) else {
        return (atoms::error(), atoms::invalid_input()).encode(env);
    };
    match find_moon_transits(
        &station(lat_deg, lon_deg, alt_km),
        start,
        end,
        step_seconds,
        time_tolerance_seconds,
    ) {
        Ok(transits) => {
            let rows: Vec<EventTerm> = transits
                .into_iter()
                .map(|transit| transit_term(env, transit))
                .collect();
            (atoms::ok(), rows).encode(env)
        }
        Err(_) => (atoms::error(), atoms::invalid_input()).encode(env),
    }
}
